src/timezones/interpret.jl

```diff
@@ -0,0 +1,193 @@
+# TimeZone concepts used to disambiguate context of DateTimes
+# abstract UTC <: TimeZone  # Defined in Base.Dates
+abstract Local <: TimeZone
+
+function transition_range(local_dt::DateTime, tz::VariableTimeZone, ::Type{Local})
+    transitions = tz.transitions
+
+    # Determine the earliest and latest possible UTC DateTime
+    # that this local DateTime could be.
+    # TODO: Alternatively we should only look at the range of offsets available within
+    # this TimeZone.
+    earliest = local_dt + MIN_OFFSET
+    latest = local_dt + MAX_OFFSET
+
+    # Determine the earliest transition the local DateTime could
+    # occur within.
+    start = searchsortedlast(
+        transitions, earliest,
+        by=el -> isa(el, Transition) ? el.utc_datetime : el,
+    )
+    start = max(start, 1)
+    finish = length(transitions)
+    for i in start:finish
+        if transitions[i].utc_datetime > latest
+            finish = i - 1
+            break
+        end
+    end
+
+    return start:finish
+end
+
+function transition_range(utc_dt::DateTime, tz::VariableTimeZone, ::Type{UTC})
+    index = searchsortedlast(
+        tz.transitions, utc_dt,
+        by=el -> isa(el, Transition) ? el.utc_datetime : el,
+    )
+    index = max(index, 1)
+    return index:index
+end
+
+"""
+    transition_range(dt::DateTime, tz::VariableTimeZone, context::Type{Union{Local,UTC}}) -> UnitRange
+
+Finds the indexes of the `tz` transitions which may be applicable for the `dt`. The given
+DateTime is expected to be local to the time zone or in UTC as specified by `context`. Note
+that UTC context will always return a range of length one.
+"""
+transition_range(dt::DateTime, tz::VariableTimeZone, context::Type{Union{Local,UTC}})
+
+function interpret(local_dt::DateTime, tz::VariableTimeZone, ::Type{Local})
+    interpretations = ZonedDateTime[]
+    t = tz.transitions
+    n = length(t)
+    for i in transition_range(local_dt, tz, Local)
+        # Convert the local DateTime into UTC
+        utc_dt = local_dt - t[i].zone.offset
+
+        if utc_dt >= t[i].utc_datetime && (i == n || utc_dt < t[i + 1].utc_datetime)
+            push!(interpretations, ZonedDateTime(utc_dt, tz, t[i].zone))
+        end
+    end
+
+    return interpretations
+end
+
+function interpret(utc_dt::DateTime, tz::VariableTimeZone, ::Type{UTC})
+    range = transition_range(utc_dt, tz, UTC)
+    length(range) == 1 || error("A UTC DateTimes should only have a single interpretation")
```

Comment:
Let the user know that this is a problem with TimeZones, not their code.

Also pick either "A UTC DateTime" or "UTC DateTimes"